algorithms/rainbow/env.py

```diff
@@ -0,0 +1,144 @@
+                    self.ale.reset_game()
+        # Process and return "initial" state
+        observation = self._get_state()
+        self.state_buffer.append(observation)
+        self.lives = self.ale.lives()
+        return torch.stack(list(self.state_buffer), 0)
+
+    def step(self, action):
+        # Repeat action 4 times, max pool over last 2 frames
+        frame_buffer = torch.zeros(2, 84, 84, device=self.device)
+        reward, done = 0, False
+        for t in range(4):
+            reward += self.ale.act(self.actions.get(action))
+            if t == 2:
+                frame_buffer[0] = self._get_state()
+            elif t == 3:
+                frame_buffer[1] = self._get_state()
+            done = self.ale.game_over()
+            if done:
+                break
+        observation = frame_buffer.max(0)[0]
+        self.state_buffer.append(observation)
+        # Detect loss of life as terminal in training mode
+        if self.training:
+            lives = self.ale.lives()
+            if 0 < lives < self.lives:  # Lives > 0 for Q*bert
+                self.life_termination = not done  # Only set flag when not truly done
+                done = True
+            self.lives = lives
+        info = None
+        # Return state, reward, done, info
+        return torch.stack(list(self.state_buffer), 0), reward, done, info
+
+    # Uses loss of life as terminal signal
+    def train(self):
+        self.training = True
+
+    # Uses standard terminal signal
+    def eval(self):
+        self.training = False
+
+    def render(self):
+        cv2.imshow('screen', self.ale.getScreenRGB()[:, :, ::-1])
+        cv2.waitKey(1)
+
+    def close(self):
+        cv2.destroyAllWindows()
+
+
+class MultipleStepsEnv(gym.Wrapper):
+    """
+    Wraps ai2thor gym environment to execute history_length steps every time its step function
+    is called
+    :param environment:
+    :return:
+    """
+    def __init__(self, env, n_steps, device):
+        gym.Wrapper.__init__(self, env)
+        self.env = env
+        self.n_steps = n_steps
+        self.device = device
+        h, w = self.env.config['resolution'][0], self.env.config['resolution'][1]
+        self.frame_buffer = torch.zeros((2, h, w), device=self.device, dtype=torch.float32)
+        self.state_buffer = deque([], maxlen=n_steps)
+
+    def step(self, action):
+        """
+        Repeat action n_step times. Regardless of the number of steps, the buffer only stores
+        the last 2 frames
+        """
+        reward, done, info = 0, False, {}
+        for t in range(self.n_steps):
+            state, reward, done, info = self.env.step(action)
+            if t == self.n_steps - 2:
+                self.frame_buffer[0] = torch.from_numpy(state)
+            elif t == self.n_steps - 1:
+                self.frame_buffer[1] = torch.from_numpy(state)
+            if done:
+                break
+        observation = self.frame_buffer.max(0)[0]
```

Comment:
Actually here is our max pool? 
im confused about this. We are actually taking the max value from each row in the 0th dimension which is of size 2 for some reason?

Reply:
is this similar to a difference image? well not really but still crazy